# Task to do :

**Preprocessed the movies_metadata dataset for the upcoming EDA** :( remove duplicates, handle missing value and mistype columns, remove outliers )

The movies dataset -> https://www.kaggle.com/rounakbanik/the-movies-dataset

This dataset is an ensemble of data collected from TMDB and GroupLens.

# 1. Loading the data

**movies_metadata.csv :** The main Movies Metadata file. Contains information on 45,000 movies featured in the Full MovieLens dataset. Features include posters, backdrops, budget, revenue, release dates, languages, production countries and companies.

>## First we import all the useful libraries 
>- Path for filepath managing
>- pandas and numpy for dataframe manip 
>- and for dataviz seaborn and matplotlib

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


>##  We organise our different files directories

In [2]:
DATA_DIR = Path('../data')
RAW_DIR = DATA_DIR / 'raw'
PROCESSED_DIR = DATA_DIR / 'processed'

>##  We load the different datasets into dataframes using Pandas

In [3]:
raw_movies_meta_df = pd.read_csv(RAW_DIR / 'movies_metadata.csv')


pd.options.display.max_columns = 50
raw_movies_meta_df.sample(3)


C:\Users\vivie\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3145: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


adult belongs_to_collection budget  \
38248  False                   NaN      0   
19585  False                   NaN      0   
34787  False                   NaN      0   

                                                  genres homepage      id  \
38248                      [{'id': 18, 'name': 'Drama'}]      NaN  391997   
19585  [{'id': 35, 'name': 'Comedy'}, {'id': 27, 'nam...      NaN   48202   
34787                      [{'id': 18, 'name': 'Drama'}]      NaN  191185   

         imdb_id original_language                            original_title  \
38248  tt0190277                en  A Body to Die For: The Aaron Henry Story   
19585  tt0054768                en             Creature from the Haunted Sea   
34787  tt2402186                pt                    Meu Pé de Laranja Lima   

                                                overview popularity  \
38248  Aaron Henry is a high school football player w...   0.605085   
19585  A crook decides to bump off members of his ine...   0.297889   
34787  Zezé is an 8 year old boy who, although a naug...   1.000202   

                            poster_path  \
38248  /j0Pib9gLimX9r5AEaqjp1sI4d5Y.jpg   
19585  /hhHW2Japfkw3Cm2rSvGtTFEAhIt.jpg   
34787  /jJYND699sY9XGZfe6om8opueWCC.jpg   

                                    production_companies  \
38248                                                 []   
19585  [{'name': 'Roger Corman Productions', 'id': 14...   
34787           [{'name': 'Passaro Films', 'id': 13967}]   

                                    production_countries release_date  \
38248                                                 []   1994-03-22   
19585  [{'iso_3166_1': 'US', 'name': 'United States o...   1961-06-01   
34787           [{'iso_3166_1': 'BR', 'name': 'Brazil'}]   2013-04-19   

       revenue  runtime                            spoken_languages    status  \
38248      0.0     30.0                                          []  Released   
19585      0.0     63.0    [{'iso_639_1': 'en', 'name': 'English'}]  Released   
34787      0.0     99.0  [{'iso_639_1': 'pt', 'name': 'Português'}]  Released   

                                                 tagline  \
38248                                                NaN   
19585  This Gangster's "Single Partner" Isn't Even Hu...   
34787                                                NaN   

                                          title  video  vote_average  \
38248  A Body to Die For: The Aaron Henry Story  False           0.0   
19585             Creature from the Haunted Sea  False           3.3   
34787                      My Sweet Orange Tree  False           6.8   

       vote_count  
38248         0.0  
19585         6.0  
34787         9.0

# 2. Preprocessing of the data 

Columns contents:

- **adult:** Indicates if the movie is X-Rated or Adult.
- **belongs_to_collection:** A stringified dictionary that gives information on the movie series the particular film belongs to.
- **budget:** The budget of the movie in dollars.
- **genres:** A stringified list of dictionaries that list out all the genres associated with the movie.
- **homepage:** The Official Homepage of the move.
- **id:** The ID of the movie.
- **imdb_id:** The identifier for the Internet Movie Database (IMDb) [with prefix 'tt', 'nm', 'co', 'ev', 'ch' or 'ni'].
- **original_language:** The language in which the movie was originally shot in.
- **original_title:** The original title of the movie.
- **overview:** A brief blurb of the movie.
- **popularity:** The Popularity Score assigned by TMDB.
- **poster_path:** The URL of the poster image.
- **production_companies:** A stringified list of production companies involved with the making of the movie.
- **production_countries:** A stringified list of countries where the movie was shot/produced in.
- **release_date:** Theatrical Release Date of the movie.
- **revenue:** The total revenue of the movie in dollars.
- **runtime:** The runtime of the movie in minutes.
- **spoken_languages:** A stringified list of spoken languages in the film.
- **status:** The status of the movie (Released, To Be Released, Announced, etc.)
- **tagline:** The tagline of the movie.
- **title:** The Official Title of the movie.
- **video:** Indicates if there is a video present of the movie with TMDB.
- **vote_average:** The average rating of the movie.
- **vote_count:** The number of votes by users, as counted by TMDB.

## 2.1. Information on the data

In [4]:
# basics info
raw_movies_meta_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [5]:
# Let us check if there is any duplicate rows
raw_movies_meta_df.duplicated().value_counts()

False    45453
True        13
dtype: int64

> The moving table:
>- has 45466 rows and 24 columns:
    >    - Some columns have null values **( we will handle thoses missing values )** 
    >    - 13 rows are duplicated so we will drop them 
>- 4 float and 20 objects 
>- amongst 20 objects :
    >    - some should be integer, float, boolean, datetime **( we will handle thoses missing types )**
    >    - some columns have information inside list of dict ( json ) **( we will extract those information )**
    

In [6]:
# To be able to see all columns
pd.options.display.max_columns = 25

## 2.2. Duplicate rows, Handling missing, incorrect and invalid data

#### 2.2.1. Duplicate rows

In [7]:
# removing the duplicate rows and create a processed_movies_meta_df
processed_movies_meta_df = raw_movies_meta_df.drop_duplicates().copy()

#### 2.2.2. missing data

In [8]:
# what is the count of missing values
count_is_null_movies_meta = raw_movies_meta_df.isnull().sum().sort_values()
count_is_null_movies_meta

adult                        0
budget                       0
genres                       0
id                           0
original_title               0
production_countries         3
production_companies         3
popularity                   5
video                        6
title                        6
spoken_languages             6
revenue                      6
vote_count                   6
vote_average                 6
original_language           11
imdb_id                     17
release_date                87
status                      87
runtime                    263
poster_path                386
overview                   954
tagline                  25054
homepage                 37684
belongs_to_collection    40972
dtype: int64

In [9]:
# Lets us display the proportion of those columns

> For The number of values under 1000 we will first see their proportion


In [10]:
# First we retrieve the columns to drop null values
null_columns_to_drop_movies_meta = count_is_null_movies_meta[count_is_null_movies_meta < 1000].index.tolist()


 just delete those values since we will still have 44466 from 45466

In [11]:
# Then we drop those null values while starting the creation of processed_movies_meta_df
processed_movies_meta_df = raw_movies_meta_df.dropna(subset = null_columns_to_drop_movies_meta).copy()
processed_movies_meta_df.isnull().sum().sort_values()

adult                        0
video                        0
title                        0
status                       0
spoken_languages             0
runtime                      0
revenue                      0
release_date                 0
production_countries         0
production_companies         0
vote_average                 0
poster_path                  0
overview                     0
original_title               0
original_language            0
imdb_id                      0
id                           0
genres                       0
budget                       0
popularity                   0
vote_count                   0
tagline                  23683
homepage                 36326
belongs_to_collection    39622
dtype: int64

> We have tagline, homepage and belongs_to_collection which has a lot of null values. 
>- Even though they could be important then are useless because of too many value lacking (23683, 36326 and 39622 ). That is more than 50%
>- A first good choice could be to drop those columns

In [12]:
processed_movies_meta_df.drop(['tagline', 'homepage', 'belongs_to_collection'], axis = 1, inplace = True ) 

#### 2.2.3. mistype

In [13]:
processed_movies_meta_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44048 entries, 0 to 45465
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   adult                 44048 non-null  object 
 1   budget                44048 non-null  object 
 2   genres                44048 non-null  object 
 3   id                    44048 non-null  object 
 4   imdb_id               44048 non-null  object 
 5   original_language     44048 non-null  object 
 6   original_title        44048 non-null  object 
 7   overview              44048 non-null  object 
 8   popularity            44048 non-null  object 
 9   poster_path           44048 non-null  object 
 10  production_companies  44048 non-null  object 
 11  production_countries  44048 non-null  object 
 12  release_date          44048 non-null  object 
 13  revenue               44048 non-null  float64
 14  runtime               44048 non-null  float64
 15  spoken_languages   

In [14]:
processed_movies_meta_df.sample(2)

adult    budget                                             genres  \
10135  False  85000000  [{'id': 35, 'name': 'Comedy'}, {'id': 14, 'nam...   
35346  False         0                     [{'id': 35, 'name': 'Comedy'}]   

          id    imdb_id original_language                     original_title  \
10135   9722  tt0374536                en                          Bewitched   
35346  63011  tt0788006                en  Assume the Position with Mr. Wuhl   

                                                overview popularity  \
10135  Thinking he can overshadow an unknown actress ...    9.27609   
35346  Emmy-winning actor-comedian Robert Wuhl assume...   0.007067   

                            poster_path  \
10135  /auUAybijSJ72XuTaPgh0EhvyJ13.jpg   
35346  /e7zxa3W387fPKaoHVsdxEqBhykS.jpg   

                                    production_companies  \
10135  [{'name': 'Columbia Pictures Corporation', 'id...   
35346                                                 []   

                                    production_countries release_date  \
10135  [{'iso_3166_1': 'US', 'name': 'United States o...   2005-06-21   
35346                                                 []   2007-10-09   

           revenue  runtime  \
10135  131426169.0    102.0   
35346          0.0     30.0   

                                        spoken_languages    status  \
10135  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
35346                                                 []  Released   

                                   title  video  vote_average  vote_count  
10135                          Bewitched  False           4.8       463.0  
35346  Assume the Position with Mr. Wuhl  False           0.0         0.0

>- some columns have inapropriate types:
>    - budget, popularity, would be better as float
>    - adult, Video would be better as boolean types
>    - release_date, runtime would be better as datetime
>- we need to extract the information inside column containing list of dict


##### a. changing dtypes of some columns

In [15]:
processed_movies_meta_df = processed_movies_meta_df.astype({'budget':'float','popularity':'float', 'video':'bool', 'release_date':'datetime64'})
processed_movies_meta_df.adult = processed_movies_meta_df.adult.apply(lambda x: not x).astype('bool')
processed_movies_meta_df.dtypes

adult                             bool
budget                         float64
genres                          object
id                              object
imdb_id                         object
original_language               object
original_title                  object
overview                        object
popularity                     float64
poster_path                     object
production_companies            object
production_countries            object
release_date            datetime64[ns]
revenue                        float64
runtime                        float64
spoken_languages                object
status                          object
title                           object
video                             bool
vote_average                   float64
vote_count                     float64
dtype: object

In [16]:
from datetime import datetime
processed_movies_meta_df['release_year'] = processed_movies_meta_df.release_date.dt.year


In [17]:
processed_movies_meta_df.runtime.describe()

count    44048.000000
mean        95.379949
std         37.212493
min          0.000000
25%         86.000000
50%         95.000000
75%        107.000000
max       1256.000000
Name: runtime, dtype: float64

In [18]:
# let us visualize a sample
processed_movies_meta_df.sample(3)

adult     budget                                             genres  \
36248  False  5000000.0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
1294   False  6000000.0  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
3371   False        0.0  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...   

           id    imdb_id original_language        original_title  \
36248  323677  tt3499096                en                  Race   
1294     9529  tt0103919                en              Candyman   
3371    85689  tt0017416                en  The Son of the Sheik   

                                                overview  popularity  \
36248  Based on the incredible true story of Jesse Ow...    7.979302   
1294   The Candyman, a murderous soul with a hook for...   12.721362   
3371   Ahmed, son of Diana and Sheik Ahmed Ben Hassan...    0.135371   

                            poster_path  \
36248  /szLfTQ2vLFQkNbP3dc0pnco1rje.jpg   
1294   /w5YQqrwzreHfa7RmXCB7rpLLxbe.jpg   
3371   /5uDB4rtXsftoXMWLSBBBFVVCvDb.jpg   

                                    production_companies  \
36248  [{'name': 'Forecast Pictures', 'id': 12686}, {...   
1294   [{'name': 'Propaganda Films', 'id': 278}, {'na...   
3371      [{'name': 'Feature Productions', 'id': 13914}]   

                                    production_countries release_date  \
36248  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   2016-02-19   
1294   [{'iso_3166_1': 'US', 'name': 'United States o...   1992-10-16   
3371   [{'iso_3166_1': 'US', 'name': 'United States o...   1926-09-05   

          revenue  runtime                                   spoken_languages  \
36248  24804129.0    134.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1294   25792310.0     99.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3371          0.0     68.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

         status                 title  video  vote_average  vote_count  \
36248  Released                  Race  False           7.0       492.0   
1294   Released              Candyman  False           6.2       303.0   
3371   Released  The Son of the Sheik  False           4.7         3.0   

       release_year  
36248          2016  
1294           1992  
3371           1926

##### b. Lets start by extracting information inside json columns ( genres, production_companies, production_countries, spoken_languages)

>##### FIRST JSON : genres

In [19]:
# Let us explore what is inside each genre columns by displaying the first 5 lines
processed_movies_meta_df.genres.tolist()[0:6]

["[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]",
 "[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]",
 "[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]",
 "[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]",
 "[{'id': 35, 'name': 'Comedy'}]",
 "[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]"]

> So we have multiple values inside each movies and each gender  :
    >- are contains in json format ( even quotes are inverted )
    >- have an id  
    >- have a name 
    
> to keep all information we will extract all name. To do so we will:
    >- extract each dict of the json
    >- retrieve the name of the gender
    >- join the name with '|' for a better display
    
> After we could explode those gender using the function explode of pandas.DataFrame:

In [20]:
import ast
def extract_name(gender):
    # The quotes says that they are string we will
    # We use the module ast to have extract the list
    gender_item = ast.literal_eval(gender)
    # We iterate inside the list to extract the name
    
    # if the name is empty we return nan
    if len(gender_item) == 0 :
        list_gender_item = np.nan
        return list_gender_item
    else:
        list_gender_item = [gender_item[elt]['name'] for elt in range(len(gender_item))] 
        # Then return the different gender join with '|'
        return '|'.join(list_gender_item)


>##### Since We could need this function in order file we will then save it in a package_module

In [21]:
# Now we apply our function to our genres
processed_movies_meta_df.genres = processed_movies_meta_df.genres.apply(extract_name)
processed_movies_meta_df.genres.sample(2)

10514    Drama|Thriller|Crime|Mystery
11158          Thriller|Crime|Foreign
Name: genres, dtype: object

In [22]:
# Now To explode 

# We need split before exploding
processed_movies_meta_df_gender_split = processed_movies_meta_df.genres.str.split('|')

processed_movies_meta_gender_explode_df = processed_movies_meta_df_gender_split.explode("genres")
processed_movies_meta_gender_explode_df.sample(2)

69695    Action
69282    Comedy
Name: genres, dtype: object

> we won't save this it to our processed_movies_meta_df for the moment later on EDA we will see if it is needed

>#### FOR THE OTHER JSON : production_companies, production_countries, spoken_languages :
>    - we will use the function extract name to have the different name 
>    - Then we won't explode them here but later if needed will do it or perhaps change them into dummies

In [23]:
processed_movies_meta_df.production_companies = processed_movies_meta_df.production_companies.apply(extract_name)


In [24]:
processed_movies_meta_df.production_countries = processed_movies_meta_df.production_countries.apply(extract_name)


In [25]:
processed_movies_meta_df.spoken_languages = processed_movies_meta_df.spoken_languages.apply(extract_name)


In [26]:
processed_movies_meta_df.sample(2)

adult  budget          genres      id    imdb_id original_language  \
272    False     0.0  Comedy|Romance    8986  tt0110516                en   
29667  False     0.0         Western  151423  tt0069962                it   

              original_title  \
272               Milk Money   
29667  Vamos a matar Sartana   

                                                overview  popularity  \
272    Three young boys pool their money and pay V, a...    3.770115   
29667  Nebraska Clay kills a gang of outlaws in self-...    0.000553   

                            poster_path  \
272    /94AgmLpD2NtCSwJ13rof6Bn5mfN.jpg   
29667  /v7e8KBo6BpuZF5uxkMoskT2aHEQ.jpg   

                                  production_companies  \
272    Paramount Pictures|The Kennedy/Marshall Company   
29667                                              NaN   

           production_countries release_date     revenue  runtime  \
272    United States of America   1994-08-31  18137661.0    108.0   
29667                       NaN   1971-09-06         0.0     87.0   

      spoken_languages    status                      title  video  \
272            English  Released                 Milk Money  False   
29667         Italiano  Released  Let's Go And Kill Sartana  False   

       vote_average  vote_count  release_year  
272             5.6        49.0          1994  
29667           0.0         0.0          1971

> We will stop here for the moment and save this processed file

In [27]:
# For convenience we will rename id column to movieId
processed_movies_meta_df = processed_movies_meta_df.rename(columns={"id": "movieId"})


In [28]:
# We will stop here and save our preprocess data
processed_movies_meta_df.to_csv(PROCESSED_DIR/ 'movies_meta.csv', index=False)

In [29]:
processed_movies_meta_df = pd.read_csv(PROCESSED_DIR/ 'movies_meta.csv')
processed_movies_meta_df.isnull().sum().sort_values()

adult                       0
vote_average                0
video                       0
title                       0
status                      0
runtime                     0
revenue                     0
release_date                0
vote_count                  0
release_year                0
popularity                  0
overview                    0
original_title              0
original_language           0
imdb_id                     0
movieId                     0
budget                      0
poster_path                 0
genres                   1966
spoken_languages         3442
production_countries     5723
production_companies    10949
dtype: int64

In [30]:
# Our extraction operation generates null value (from '[]' )
# Since spoken language should have and least the original_language as spoken we can fill with this value
# Same for production_countries
# But because of the low proportion, we will first drop those null values
processed_movies_meta_df = processed_movies_meta_df.dropna(axis = 0 ) 

In [31]:
# We will stop here and save our preprocess data
processed_movies_meta_df.to_csv(PROCESSED_DIR/ 'movies_meta.csv', index=False)